src/NServiceBus.Gateway/Gateway.cs

```diff
@@ -125,20 +148,105 @@ static void ConfigureReceiver(FeatureConfigurationContext context, Address gatew
+        {
+            public GatewayReceiverStartupTask(IManageReceiveChannels channelManager, IRouteMessagesToEndpoints endpointRouter, IDispatchMessages dispatcher, Func<IReceiveMessagesFromSites> receiveMessagesFromSitesFactory, string replyToAddress)
+            {
+                dispatchMessages = dispatcher;
+                routeMessagesToEndpoints = endpointRouter;
+                this.receiveMessagesFromSitesFactory = receiveMessagesFromSitesFactory;
+                manageReceiveChannels = channelManager;
+                this.replyToAddress = replyToAddress;
+            }
+            
+            protected override Task OnStart(IBusSession context)
+            {
+                foreach (var receiveChannel in manageReceiveChannels.GetReceiveChannels())
+                {
+                    var receiver = receiveMessagesFromSitesFactory();
+
+                    receiver.Start(receiveChannel, receiveChannel.NumberOfWorkerThreads, MessageReceivedOnChannel);
+                    activeReceivers.Add(receiver);
+
+                    Logger.InfoFormat("Receive channel started: {0}", receiveChannel);
+                }
+
+                return Task.FromResult(0);
+            }
+
+            protected override Task OnStop(IBusSession context)
+            {
+                Logger.InfoFormat("Receiver is shutting down");
+
+                foreach (var channelReceiver in activeReceivers)
+                {
+                    Logger.InfoFormat("Stopping channel - {0}", channelReceiver.GetType());
+
+                    channelReceiver.Dispose();
+                }
+
+                activeReceivers.Clear();
+
+                Logger.InfoFormat("Receiver shutdown complete");
+                return Task.FromResult(0);
+            }
+
+            async Task MessageReceivedOnChannel(MessageReceivedOnChannelArgs e)
+            {
+                var body = e.Body;
+                var headers = e.Headers;
+                var id = e.Id;
+                var recoverable = e.Recoverable;
+                var timeToBeReceived = e.TimeToBeReceived;
+
+                var destination = routeMessagesToEndpoints.GetDestinationFor(headers);
+
+                Logger.Info("Sending message to " + destination);
+
+                var outgoingMessage = new OutgoingMessage(id, headers, body);
+                outgoingMessage.Headers[Headers.ReplyToAddress] = replyToAddress;
+
+                var deliveryConstraints = new List<DeliveryConstraint>
+                {
+                    new DiscardIfNotReceivedBefore(timeToBeReceived)
+                };
+
+                if(!recoverable)
+                {
+                    deliveryConstraints.Add(new NonDurableDelivery());
+                }
+
+                var operation = new UnicastTransportOperation(outgoingMessage, destination, deliveryConstraints);
+                await dispatchMessages.Dispatch(WrapInOperations(operation), new ContextBag()).ConfigureAwait(false);
```

Comment:
This will be affected by https://github.com/Particular/NServiceBus/pull/3357. I'll send in a PR